File: java/src/main/java/org/sparky/model/Table.java
package org.sparky.model;

import org.sparky.model.exceptions.InvalidColumnNameException;
import org.sparky.model.exceptions.InvalidKeyException;

import java.util.*;

/**
 * Implementation of a table
 *
 * Created by lloyd on 19/10/2015.
 */
public class Table {
    public static final String INDEX_KEY = "$$";

    private Map<String, Integer> columns = new HashMap<>();
    private String[][] data;
    private Map<String, Filter> filters = new HashMap<>();

    private String[][] cachedFiltered;
    private Integer currentRow = null;

    public Table(Map<String, Integer> columns, String[][] data, Map<String, Filter> filters) {
        this.columns = columns;
        this.data = data;
        this.filters = filters;
    }

    /**
     * Indexes Run 1..n not 0..n
     * @param tableIndex
     * @param root
     * @return
     * @throws InvalidKeyException
     */
    public boolean isIndexValid(Integer tableIndex, Configuration root) throws InvalidKeyException {
        if(haveFiltersChanged(root) || cachedFiltered == null){
            applyFilters(root);
        }

        return ((tableIndex-1) <= cachedFiltered.length && tableIndex >= 0);
    }

    public void setRowIndex(Integer rowIndex){
        currentRow = rowIndex-1;
    }

    public void clearRowIndex(){
        currentRow = null;
    }

    public List<String> valuesForColumn(String columnName, Configuration root) throws InvalidKeyException {
        if(haveFiltersChanged(root) || cachedFiltered == null){
            applyFilters(root);
        }

        int firstRow = currentRow == null ? 0 : currentRow;
        int maxRow = currentRow == null ? cachedFiltered.length : currentRow + 1;

        if(columnName.equals(INDEX_KEY)){
            List<String> indexes = new ArrayList<>();
            for(int row = firstRow; row < maxRow; row++){
                indexes.add(Integer.toString(row+1));
            }
            return indexes;
        }

        if(columns.containsKey(columnName)){
            int columnIndex = columns.get(columnName);

            Set<String> values = new HashSet<>();

            for(int row = firstRow; row < maxRow; row++){
                values.add(cachedFiltered[row][columnIndex]);
            }

            return new ArrayList<>(values);
        } else {
            throw new InvalidColumnNameException(columnName);
        }
    }

    public String valueForColumn(String columnName, Configuration root) throws InvalidKeyException {
        List<String> values = valuesForColumn(columnName, root);

        if(values.size() == 1){
            return values.get(0);
        }
        throw new InvalidKeyException(String.format("Query for '%s' Column returned either 0 or > 1 value.", columnName));
    }

    private void applyFilters(Configuration root) throws InvalidKeyException {
        Map<Integer, String> evaluatedFilters = new HashMap<>();
        for(String col : filters.keySet()){
            Filter fit = filters.get(col);
            Integer colIndex = columns.get(col);
            evaluatedFilters.put(colIndex, fit.getValue(root));
        }

        if(evaluatedFilters.size() == 0){
            cachedFiltered = data;
            return;
        }

        List<String[]> filteredRows = new ArrayList<>();
        for(int row = 0; row < data.length; row++){
            String[] values = data[row];
            boolean allowRow = true;
            for(Integer index : evaluatedFilters.keySet()){
                if(!values[index].equals(evaluatedFilters.get(index))){
                    allowRow = false;
                    break;
                }
            }
            if(allowRow){
                filteredRows.add(values);
            }
        }
        cachedFiltered = new String[filteredRows.size()][];
        filteredRows.toArray(cachedFiltered);
    }

    private boolean haveFiltersChanged(Configuration root) throws InvalidKeyException {
        for(Filter f : filters.values()){
            if(f.hasChanged(root)){
                return true;
            }
        }
        return false;
    }
}

File: net/Sparky/Model/Table.cs
using Sparky.Model.Exceptions;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Sparky.Model
{
    public class Table
    {
        public static readonly string INDEX_KEY = "$$";

        private IDictionary<string, int> _columns;
        private string[][] _data;
        private IDictionary<string, Filter> _filters;

        private string[][] _cachedFilteredData;
        private int _rowIndexLock = -1;

        public Table(IDictionary<string, int> columns, string[][] data, IDictionary<string, Filter> filters)
        {
            _columns = columns;
            _data = data;
            _filters = filters;
        }

        internal string ValueForColumn(string columnName, IConfiguration root)
        {
            var values = ValuesForColumn(columnName, root);
            if(values.Count == 1)
            {
                return values.First();
            }
            return null;
        }

        internal ICollection<string> ValuesForColumn(string columnName, IConfiguration root)
        {
            if (HaveFiltersChanged(root) || _cachedFilteredData == null)
            {
                ApplyFilters(root);
            }

            int firstRow = _rowIndexLock == -1 ? 0 : _rowIndexLock;
            int lastRow = _rowIndexLock == -1 ? _cachedFilteredData.Length : _rowIndexLock + 1;

            if(columnName == INDEX_KEY)
            {
                var indexes = new List<string>();
                for(int r = firstRow; r < lastRow; r++)
                {
                    indexes.Add((r + 1).ToString());
                }
                return indexes;
            }

            if (_columns.ContainsKey(columnName))
            {
                int colInd = _columns[columnName];
                var values = new HashSet<string>();

                for (int r = firstRow; r < lastRow; r++)
                {
                    values.Add(_cachedFilteredData[r][colInd]);
                }
                return values;
            }

            throw new InvalidColumnNameException(columnName);
        }

        internal bool IsIndexValid(int index, IConfiguration root)
        {
            if (HaveFiltersChanged(root) || _cachedFilteredData == null)
            {
                ApplyFilters(root);
            }

            return ((index - 1) <= _cachedFilteredData.Length && index > 0);
        }

        private void ApplyFilters(IConfiguration root)
        {
            var evaluatedFilter = new Dictionary<int, string>();
            foreach(var col in _filters.Keys)
            {
                var f = _filters[col];
                var colIndex = _columns[col];
                evaluatedFilter.Add(colIndex, f.Value(root));
            }

            if(evaluatedFilter.Count == 0)
            {
                _cachedFilteredData = _data;
            } 
            else
            {
                List<string[]> filteredRows = new List<string[]>();
                for(int r = 0; r < _data.Length; r++)
                {
                    var values = _data[r];
                    var allowRow = true;
                    foreach(var colInd in evaluatedFilter.Keys)
                    {
                        if(values[colInd] != evaluatedFilter[colInd])
                        {
                            allowRow = false;
                            break;
                        }
                    }
                    if (allowRow)
                    {
                        filteredRows.Add(values);
                    }
                }
                _cachedFilteredData = filteredRows.ToArray();
            }
        }

        private bool HaveFiltersChanged(IConfiguration root)
        {
            foreach(Filter f in _filters.Values)
            {
                if (f.HasChanged(root))
                {
                    return true;
                }
            }
            return false;
        }

        internal void SetRowIndex(int index)
        {
            _rowIndexLock = index-1;
        }

        internal void ClearRowIndex()
        {
            _rowIndexLock = -1;
        }
    }
}
